### PR TITLE
Fix NRE introduced in FileSystemWatcher during nullability annotations

### PR DESCRIPTION
**Pull request by krwq**
While annotating https://github.com/dotnet/runtime/pull/41261 I've introduced a product bug which luckily got caught by tests. It is weird compiler behavior (https://github.com/dotnet/csharplang/issues/3393) which changes `?.` precedence when adding `!`, see i.e.:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAoEvAZgAIBvEy+y7fANkYBZKAFCGAC3wAUAYXyUADgEoadBrIBuEBJQB2EMHEoBecQH4AdENx6AcmrgBuGbPpMAnANXqJl4rIC+Vyp6atsHbny4wqKS0q7W9ApKjhraYvqGAIQmZi4RNvj2Mc6eHsR5ZLiUIiS04TZUhsVF1JQA5nAw5pRQjc15BcR41aXeVEwADJSm6jT1bS1tHUA

I've done quick search in our code if we've introduced similar errors elsewhere and found only this.

Note: **this is fixing a regression** introduced by https://github.com/dotnet/runtime/commit/e59c9260eb78bbf093d8081a3b8a1ecc0132d70f#diff-ab050074db1f148b9d552c569a7899b7R398.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Ouch. Nasty compiler bug. Can you add a test?

**Comment by danmosemsft** [comment]
> I've done quick search in our code if we've introduced similar errors elsewhere and found only this.

How did you do such a search? Look for "?." followed by "!." ? I am wondering how reliable a check this is and whether we ought to ask for an instrumented compiler to help find this cas.e

File: src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs [hunk 1]
```diff
@@ -390,7 +390,11 @@ private void AddDirectoryWatchUnlocked(WatchedDirectory? parent, string director
                     // of the world, but there's little that can be done about that.)
                     if (directoryEntry.Parent != parent)
                     {
-                        directoryEntry.Parent?.Children!.Remove (directoryEntry);
+                        if (directoryEntry.Parent != null)
```
Comment (jeffhandley) [inline src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs:393]:
```suggestion
                        // Work around https://github.com/dotnet/csharplang/issues/3393 preventing Parent?.Children!. from behaving as expected
                        if (directoryEntry.Parent != null)
```

File: src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs [hunk 2]
```diff
@@ -443,7 +447,11 @@ private void RemoveWatchedDirectory(WatchedDirectory directoryEntry, bool remove
                 Debug.Assert (_includeSubdirectories);
                 lock (SyncObj)
                 {
-                    directoryEntry.Parent?.Children!.Remove(directoryEntry);
+                    if (directoryEntry.Parent != null)
```
Comment (jeffhandley) [inline src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs:450]:
```suggestion
                    // Work around https://github.com/dotnet/csharplang/issues/3393 preventing Parent?.Children!. from behaving as expected
                    if (directoryEntry.Parent != null)
```

**Review by jeffhandley — APPROVED** [review]
Please apply code comments similar to my suggestions; otherwise looks good to me.

**Comment by krwq** [comment]
@danmosemsft

> How did you do such a search?

RegEx, yes `?.` followed by `!.` (only single line searches), perhaps we could do so but IMO there will be relatively small chance of that happening (I was expecting more occurrences with RegExes), this would be ideal if we have cycles for that - I personally am not too familiar with how to search for that using Roslyn and also expecting there will be no more entries but if someone knows how to do it quickly I recommend we should do it (perhaps even better might be to ask compiler team to fix this, especially I've seen suggested code fix in one of the compiler issues).

> Can you add a test?

I'll see if I can find some easy way to repro but considering this is affecting Linux only and possibly some edge case I'm not sure how hard this will be so I'll time box the research.

File: src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs (hunk 1 above)
Comment (krwq) [reply src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs:393]: I initially thought to add it but then I thought that we introduced it while annotating by doing tiny refactoring so wasn't sure about adding it. Will change

**Comment by danmosemsft** [comment]
@agocke any suggestions for how we might find any other instances of this insidious error? Beyond the regex for `\?\.\w+!\.` which @krwq already performed

**Comment by krwq** [comment]
@danmosemsft I actually search twice per line so I don't miss cases like `foo?.bar.baz!.foo2;` Didn't want to mess with Regexes too much as it's easy to make stupid mistake.

<details>
<summary>Here is the code I used</summary>

```csharp
        static readonly Regex s_r1 = new Regex(@"[a-z0-9_]\?\.", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        static readonly Regex s_r2 = new Regex(@"[a-z0-9_]\!\.", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        static bool IsBuggyCompilerCase(string line)
        {
            Match m1 = s_r1.Match(line);
            if (m1.Success)
            {
                int start = m1.Index;
                foreach (Match m2 in s_r2.Matches(line))
                {
                    int end = m2.Index;
                    if (start < end)
                    {
                        return true;
                    }
                }
            }

            return false;
        }
```

+ searching in all directories for all .cs files

</details>

**Comment by krwq** [comment]
@danmosemsft I've tried several ways to repro this including with symlinks and other combinations of removing/renaming (as suggested per comment) but I'm unsure how to do it or if it's even reproducible. Probably owners might have some clues or ideas (cc: @maryamariyan @carlossanlop). Looking at the code it's related to existing file handle which doesn't have a parent which suggests a root of FS Watcher - I've also tried removing the root but also haven't seen anything which would manifest as any sort of NRE.. I'd still do the fix just in case but not 100% sure if worth porting unless we understand what might possibly be broken

**Review by carlossanlop — APPROVED** [review]

**Comment by carlossanlop** [comment]
Does this need to be backported anywhere?

**Comment by carlossanlop** [comment]
@krwq can you try this regex?

```
\w(\([\w\s,]*\))?[!?]\.+(\w(\([\w\s,]*\))?[!?]\.)+
```

It's not as thorough as Roslyn could be, but at least it can find a few more instances. For example, it worked with this code:

```cs
    static void Main()
    {
        A?.B.D; // No
        A!.B.C!.D; // No
        A()!.B().C()!.D; // No
        A(a, b)!.B(b, c).C(c, d)?.D; // No
        A?.B!.C; // Yes
        A?.B!.C!.D; // Yes
        A!.B?.C!.D?.E; // Yes
        A()!.B()?.C().D; // Yes
        A(a)!.B(b)?.C().D; // Yes
        A(a, b)!.B(b, c)!.C.D; // Yes
    }
```
Please correct me if any of the cases is wrong.

**Review by danmoseley — APPROVED** [review]

**Comment by danmosemsft** [comment]
It's OK that you can't figure out how to test it. That happens. The change clearly returns us to previous behavior, and your sharplab.io demonstrates why we need to do that. Please port into 5.0.

Any further regexing throws up another instance, please do follupw fix.

**Comment by danmosemsft** [comment]
Oh I can trigger it

/backport to release/5.0

**Comment by danmosemsft** [comment]
Oops.. let me close one..

**Comment by danmosemsft** [comment]
@akoeplinger just curious, how did the bot not end up creating two PR's here?

**Comment by akoeplinger** [comment]
@danmosemsft before opening a PR it checks whether the branch already exists on the origin remote to avoid opening another PR. This allows you to push changes to the source PR and do another `/backport` and the changes will be pushed to the existing backport branch/PR.

**Comment by krwq** [comment]
@carlossanlop I wasn't able to find anything new with your Regex directly but your tests gave me idea of one important case I missed (calls in the chain) and I found one more.

**Comment by carlossanlop** [comment]
@danmosemsft is [your comment](https://github.com/dotnet/runtime/pull/41315#issuecomment-680171381) going to do the port automatically? I ask because I have [this PR open](https://github.com/dotnet/runtime/pull/41348) to port all the nullability changes manually together, so I'd like to know if I should skip this one.

**Comment by carlossanlop** [comment]
@danmosemsft confirmed, this change is already merged to release/5.0: https://github.com/dotnet/runtime/commit/2aa627b7cf8cc1c697f916c7cd6e76fa52b69576

I won't include it in my PR.

**Pull request merged**